### PR TITLE
add all_vault_utilizations query to CM

### PR DESCRIPTION
**Pull request by bobthebuidlr**

### CODE REVIEW

File: contracts/credit-manager/src/query.rs [hunk 1]
```diff
@@ -142,12 +142,50 @@ pub fn query_vault_utilization(
...
+    env: Env,
+    start_after: Option<String>,
+    limit: Option<u32>,
+) -> ContractResult<Vec<VaultUtilizationResponse>> {
+    let params = PARAMS.load(deps.storage)?;
+    let oracle = ORACLE.load(deps.storage)?;
+    let vault_configs = params.query_all_vault_configs(&deps.querier, start_after, limit)?;
+
+    vault_configs
+        .iter()
```
Comment (piobab) [inline contracts/credit-manager/src/query.rs:171]: If you change order you will avoid addr clone: 
```
Ok(VaultUtilizationResponse {
                utilization: vault_utilization_in_deposit_cap_denom(
                    &deps,
                    &oracle,
                    &vault_config,
                    &env.contract.address,
                )?,
                vault: Vault {
                    address: vault_config.addr,
                }
                .into(),
            })
```

File: contracts/credit-manager/src/query.rs [hunk 2]
```diff
@@ -142,12 +142,50 @@ pub fn query_vault_utilization(
...
+        utilization: vault_utilization_in_deposit_cap_denom(
+            &deps,
+            &oracle,
+            &vault_config,
+            &env.contract.address,
+        )?,
     })
 }
 
+pub fn query_all_vault_utilizations(
```
Comment (piobab) [inline contracts/credit-manager/src/query.rs:160]: Can we use new pagination?

File: contracts/credit-manager/src/query.rs (hunk 1 above)
Comment (bobthebuidlr) [reply contracts/credit-manager/src/query.rs:171]: Why is this supposed to make the `clone()` redundant? I just tried and it still not allowing moving of borrowed `addr`.

File: contracts/credit-manager/src/query.rs (hunk 2 above)
Comment (bobthebuidlr) [reply contracts/credit-manager/src/query.rs:160]: For that to work, I'd have to adjust the `query_all_vault_configs` as well to support the pagination. This means we'd also have to create a v2 version for that one. Should I still do this? Wouldn't hurt I guess, but will just take some more time.

Comment (piobab) [reply contracts/credit-manager/src/query.rs:160]: Up to you. If we create new queries IMO we should go with newest format which will be end goal.

Comment (bobthebuidlr) [reply contracts/credit-manager/src/query.rs:160]: Yeah I do agree here. Let me just add another V12 query for vault_configs to support the new pagination then.

File: packages/types/tests/test_paginate_query.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#[test]
```
Comment (bobthebuidlr) [inline packages/types/tests/test_paginate_query.rs:1]: @piobab I'm unsure how to create a mock `Map` that we can use for testing this query

File: packages/types/src/lib.rs [hunk 4]
```diff
@@ -26,3 +30,33 @@ pub struct PaginationResponse<T> {
 pub struct Metadata {
     pub has_more: bool,
 }
+
+pub fn paginate_query<'a, K, T, R, E, F>(
```
Comment (bobthebuidlr) [inline packages/types/src/lib.rs:34]: This one only works with `Map`. For sure RB (and probably other contracts) will need to use a `prefix`. Should we just create a second version of this function for prefix variants (feels very duplicate IMO, should be a better way)? @piobab

Comment (piobab) [reply packages/types/src/lib.rs:34]: Could be tricky to create correct generic types to handle prefix and map in the same function. For now let's do second function as it is in cw-paginate.

File: packages/types/tests/test_paginate_query.rs (hunk 3 above)
Comment (piobab) [reply packages/types/tests/test_paginate_query.rs:1]: You can do something like this:
```
#[test]
fn sample_test_case() {
    let mut deps = mock_dependencies();
    let env = mock_env();

    let coin_balances: Map<&str, Uint128> = Map::new("coin_balance");

    coin_balances.save(&mut deps.storage, "key_1", &Uint128::new(100)).unwrap();
    coin_balances.save(&mut deps.storage, "key_2", &Uint128::new(101)).unwrap();
    coin_balances.save(&mut deps.storage, "key_3", &Uint128::new(102)).unwrap();

    let res = paginate_query(&coin_balances, &deps.storage, None, 2, |key, amount| {
        let mapped = CoinBalanceResponseItem {
            account_id: "accoun_1".to_string(),
            denom: key,
            amount,
        };

        Ok::<CoinBalanceResponseItem, StdError>(mapped)
    });
    let res = res.unwrap();
    assert_eq!(res.metadata.has_more, true);
    assert_eq!(res.data.len(), 2);
}
```

File: contracts/credit-manager/tests/tests/test_utilizations_all_query.rs [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+#[test]
+fn utilizations_are_zero() {
+    let vault_info_1 = unlocked_vault_info();
+    let vault_info_2 = locked_vault_info();
+
+    let mock = MockEnv::new().vault_configs(&[vault_info_1.clone(), vault_info_2]).build().unwrap();
+
+    let res = mock.query_all_vault_utilizations(None, None).unwrap();
+
+    let _ = res.data.iter().map(|vault_utilization| {
```
Comment (piobab) [inline contracts/credit-manager/tests/tests/test_utilizations_all_query.rs:40]: `res.data.iter()` could be used even if in the response are only one item. It doesn't check if there are 2 items.

I would prefer to match like this:
```
assert_eq!(
        res.data,
        vec![
            VaultUtilizationResponse {
                vault: mock.get_vault(&vault_info_1),
                utilization: Coin {
                    denom: vault_info_1.deposit_cap.denom.clone(),
                    amount: Uint128::zero(),
                },
            },
            VaultUtilizationResponse {
                vault: mock.get_vault(&vault_info_2),
                utilization: Coin {
                    denom: vault_info_2.deposit_cap.denom.clone(),
                    amount: Uint128::zero(),
                },
            }
        ]
    );
```

File: contracts/params/tests/tests/test_query_all_vault_configs_v2.rs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        &mock.query_owner(),
+        VaultConfigUpdate::AddOrUpdate {
+            config: vault_config_2,
+        },
+    )
+    .unwrap();
+
+    let vault_configs = mock.query_all_vault_configs_v2(None, Some(1));
+
+    assert_eq!(vault_configs.data.get(0).unwrap().addr, deps.api.addr_validate("vault_1").unwrap())
```
Comment (piobab) [inline contracts/params/tests/tests/test_query_all_vault_configs_v2.rs:31]: You can use:
`Addr::unchecked("vault_1")`

**Review by piobab — APPROVED** [review]
LGTM. Only minor changes

File: contracts/params/tests/tests/test_query_all_vault_configs_v2.rs (hunk 6 above)
Comment (piobab) [inline contracts/params/tests/tests/test_query_all_vault_configs_v2.rs:31]: I will check length also

File: contracts/params/tests/tests/test_query_all_vault_configs_v2.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    .unwrap();
+    mock.update_vault_config(
+        &mock.query_owner(),
+        VaultConfigUpdate::AddOrUpdate {
+            config: vault_config_2,
+        },
+    )
+    .unwrap();
+
+    let vault_configs = mock.query_all_vault_configs_v2(None, Some(1));
```
Comment (piobab) [inline contracts/params/tests/tests/test_query_all_vault_configs_v2.rs:55]: Would be nice to check that `Bound::exclusive` works

Comment (bobthebuidlr) [reply contracts/params/tests/tests/test_query_all_vault_configs_v2.rs:55]: Done. Also added an extra test case for that.

File: contracts/credit-manager/tests/tests/test_utilizations_all_query.rs (hunk 5 above)
Comment (bobthebuidlr) [reply contracts/credit-manager/tests/tests/test_utilizations_all_query.rs]: Agree, adjusted accordingly

**Pull request merged**